// webpack-info-plugin/lib/WebpackInfoPlugin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var WebpackInfoPlugin = function () {
  function WebpackInfoPlugin() {
    var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, WebpackInfoPlugin);

    this.options = options;
    // the state, false: bundle invalid, true: bundle valid
    this.state = true;
  }

  _createClass(WebpackInfoPlugin, [{
    key: 'apply',
    value: function apply(compiler) {
      var _this = this;

      compiler.plugin('done', function (stats) {
        _this.showStats(stats);
        _this.showValidState(stats);
      });

      compiler.plugin('failed', function (err) {
        _this.handleError(err);
      });

      compiler.plugin('invalid', function () {
        _this.showInvalidState(false);
      });

      compiler.plugin('watch-run', function (c, callback) {
        _this.showInvalidState(false);
        callback();
      });

      compiler.plugin('run', function (c, callback) {
        _this.showInvalidState(false);
        callback();
      });
    }
  }, {
    key: 'isShowStats',
    value: function isShowStats() {
      return !!this.options.stats;
    }
  }, {
    key: 'isShowState',
    value: function isShowState() {
      return !!this.options.state;
    }
  }, {
    key: 'showStats',
    value: function showStats(stats) {
      if (this.isShowStats()) {
        console.log(stats.toString(this.options.stats)); // eslint-disable-line no-console
      }
    }
  }, {
    key: 'showInvalidState',
    value: function showInvalidState() {
      if (this.isShowState() && this.state) {
        console.log('webpack: bundle is now ' + _chalk2.default.red('INVALID') + '.'); // eslint-disable-line no-console,max-len
        this.state = false;
      }
    }
  }, {
    key: 'showValidState',
    value: function showValidState(stats) {
      if (this.isShowState() && !this.state && !stats.hasErrors() && !stats.hasWarnings()) {
        console.log('webpack: bundle is now ' + _chalk2.default.green('VALID') + '.'); // eslint-disable-line no-console,max-len
        this.state = true;
      }
    }
  }]);

  return WebpackInfoPlugin;
}();

exports.default = WebpackInfoPlugin;
module.exports = exports['default'];
